File: src/net/tsv/TsvWriter.cs
// FormatFactory.Tsv — TSV Writer
// commercial_product_ready: false

using System.Text;

namespace FormatFactory.Tsv;

/// <summary>
/// Writes tab-separated values to strings or files.
///
/// Behaviour:
///   - Joins fields with tab character, rows with LF.
///   - Null fields are treated as empty strings.
///   - Validates that fields do not contain tab or newline characters.
///   - Outputs UTF-8 without BOM.
///
/// MWP status: minimal viable product.
/// </summary>
public static class TsvWriter
{
    /// <summary>
    /// Serialize rows to a TSV string.
    /// </summary>
    /// <param name="rows">Rows of fields. Null fields are treated as empty strings.</param>
    /// <returns>TSV content as a string with LF line endings.</returns>
    /// <exception cref="TsvException">Thrown when a field contains a tab or newline character.</exception>
    public static string WriteRows(IEnumerable<IEnumerable<string?>> rows)
    {
        ArgumentNullException.ThrowIfNull(rows);

        var sb = new StringBuilder();
        foreach (var row in rows)
        {
            var fields = new List<string>();
            foreach (var field in row)
            {
                var value = field ?? string.Empty;
                ValidateField(value);
                fields.Add(value);
            }
            sb.Append(string.Join("\t", fields));
            sb.Append('\n');
        }
        return sb.ToString();
    }

    /// <summary>
    /// Serialize rows and write to a file. UTF-8, no BOM.
    /// Creates parent directories as needed.
    /// </summary>
    public static void WriteRowsToFile(IEnumerable<IEnumerable<string?>> rows, string path)
    {
        ArgumentNullException.ThrowIfNull(rows);
        if (string.IsNullOrWhiteSpace(path))
            throw new TsvException("path must not be null or empty.");

        var dir = Path.GetDirectoryName(path);
        if (!string.IsNullOrEmpty(dir)) Directory.CreateDirectory(dir);

        var content = WriteRows(rows);
        File.WriteAllText(path, content, new UTF8Encoding(encoderShouldEmitUTF8Identifier: false));
    }

    // -------------------------------------------------------------------------
    // Validation
    // -------------------------------------------------------------------------

    private static void ValidateField(string value)
    {
        if (value.Contains('\t'))
            throw new TsvException($"Field contains a tab character which is not allowed in TSV fields: \"{Truncate(value)}\"");
        if (value.Contains('\n') || value.Contains('\r'))
            throw new TsvException($"Field contains a newline character which is not allowed in TSV fields: \"{Truncate(value)}\"");
    }

    private static string Truncate(string value, int maxLength = 50)
    {
        if (value.Length <= maxLength) return value;
        return value[..maxLength] + "...";
    }
}

File: src/python/tsv/tsv_writer.py
"""
tsv_writer.py — TSV writer for format-factory-tsv.

Public API:
  write_tsv(rows, path, headers=None) — writes TSV to file
  write_tsv_str(rows, headers=None)   — returns TSV string

TSV convention: tabs within field values are illegal. This writer
raises TsvWriteError if any field value contains a tab character.
"""

from __future__ import annotations

from pathlib import Path
from .exceptions import TsvError, TsvParseError, TsvWriteError


def _format_field(value: object) -> str:
    s = "" if value is None else str(value)
    if "\t" in s:
        raise TsvWriteError(f"TSV field value must not contain tab characters: {s!r}")
    return s


def write_tsv_str(
    rows: list[list[object]],
    headers: list[str] | None = None,
) -> str:
    """Serialize rows to TSV string.

    Args:
        rows: Data rows, each a list of field values.
        headers: Optional header row prepended to output.

    Returns:
        TSV string with LF line endings.
    """
    if rows is None:
        raise TsvWriteError("rows must not be None")

    lines: list[str] = []
    if headers is not None:
        lines.append("\t".join(_format_field(h) for h in headers))

    for row in rows:
        lines.append("\t".join(_format_field(f) for f in row))

    return "\n".join(lines) + "\n" if lines else ""


def write_tsv(
    rows: list[list[object]],
    path: str | Path,
    headers: list[str] | None = None,
) -> None:
    """Write TSV rows to a file. Creates parent dirs as needed. UTF-8, no BOM."""
    if not path:
        raise TsvWriteError("path must not be empty")
    p = Path(path)
    p.parent.mkdir(parents=True, exist_ok=True)
    content = write_tsv_str(rows, headers=headers)
    p.write_text(content, encoding="utf-8")
